rush-plugins/rush-operation-resource-plugin/src/RushOperationResourcePlugin.ts

```diff
@@ -0,0 +1,214 @@
+  IRushPlugin,
+  IPhasedCommand,
+  RushSession,
+  RushConfiguration,
+  ILogger,
+  IOperationRunnerContext,
+  IOperationExecutionResult,
+  OperationStatus,
+  Operation
+} from '@rushstack/rush-sdk';
+import { Stopwatch } from '@rushstack/rush-sdk/lib/utilities/Stopwatch';
+import * as fs from 'fs';
+
+export interface IRushOperationResourcePluginOptions {
+  resourceConstraints?: IResourceConstraintConfig[];
+  resourcePools?: IResourcePoolConfig[];
+}
+
+export interface IResourceConstraintConfig {
+  appliesTo: {
+    operationName?: string;
+    phaseName?: string;
+    packageName?: string;
+    tagName?: string;
+  };
+  resourcePool: {
+    poolName: string;
+    envVarName?: string;
+  };
+}
+
+export interface IResourcePoolConfig {
+  poolName: string;
+  resources?: string[];
+  resourceCount?: number;
+}
+
+export interface IResourcePool {
+  poolName: string;
+  resources: string[];
+}
+
+export const RESOURCE_POOL_POLL_INTERVAL_MS = 500;
+export const PLUGIN_NAME: string = 'RushOperationResourcePlugin';
+
+export class RushOperationResourcePlugin implements IRushPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+  private readonly _options: IRushOperationResourcePluginOptions;
+  private readonly _checkedOutResources: Map<Operation, string>;
+  private readonly _resourcePools: Map<string, IResourcePool>;
+
+  public constructor(options: IRushOperationResourcePluginOptions) {
+    this._options = options;
+    this._checkedOutResources = new Map();
+    this._resourcePools = new Map();
+  }
+
+  public apply(rushSession: RushSession, rushContext: RushConfiguration): void {
+    const logger: ILogger = rushSession.getLogger(this.pluginName);
+
+    this._validateOptions();
+
+    rushSession.hooks.runAnyPhasedCommand.tapPromise(
+      this.pluginName,
+      async (command: IPhasedCommand): Promise<void> => {
+        command.hooks.beforeExecuteOperation.tapPromise(
+          this.pluginName,
+          async (
+            runnerContext: IOperationRunnerContext & IOperationExecutionResult
+          ): Promise<OperationStatus | undefined> => {
+            const constraint: IResourceConstraintConfig | undefined = this._getConstraintsForOperation(
+              runnerContext.operation
+            );
+
+            if (constraint) {
+              if (constraint.resourcePool) {
+                const pool: IResourcePool = this._resourcePools.get(constraint.resourcePool.poolName)!;
+
+                // Wait to obtain a resource from the resource pool for this constraint
+                const resource: string = await this._obtainResource(pool, runnerContext.operation);
+
+                // Pass this resource to the underlying operation using the specified
+                // environment variable
+                if (constraint.resourcePool.envVarName) {
+                  const runner: any = runnerContext.operation.runner;
+                  if (!runner.environment) {
+                    runner.environment = {};
```

Comment:
Is `runner.environment` already part of the specification for `IOperationRunner`? If not, I can see a number of uses for it; for example I've been playing a lot with cpu profiling lately and being able to customize `NODE_OPTIONS` per operation via plugin would be handy.

That said, probably safer to have environment be calculated via a dedicated `getEnvironmentForOperation` hook that multiple plugins can tap and thereby augment.

Reply:
Makes sense -- as the `:any` probably gives away, this part is a TOTAL hack, I'm not even checking right now (although I'm assuming) that it is a ScriptOperation haha.